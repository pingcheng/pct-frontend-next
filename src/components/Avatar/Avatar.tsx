import Image from "next/image";
import { profile } from "@/data/profile";

type AvatarProps = {
  width: number;
  height: number;
};

export function Avatar({ width, height }: AvatarProps) {
  return (
    <div
      className="max-w-full"
      style={{
        width,
        height,
      }}
    >
      <Image
        src={profile.avatarUrl}
        alt={profile.fullName}
        width={width}
        height={height}
        style={{
          borderRadius: "50%",
        }}
      />
    </div>
  );
}
