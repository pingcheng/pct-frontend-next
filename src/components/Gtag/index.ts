import Gtag from "./Gtag";

export default Gtag;
