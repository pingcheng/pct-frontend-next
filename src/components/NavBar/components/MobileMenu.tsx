"use client";

import Link from "next/link";
import { usePathname } from "next/navigation";
import styles from "../style.module.css";
import { forwardRef } from "react";

type MenuItem = {
  label: string;
  path: string;
  activePattern?: RegExp;
};

type MobileMenuProps = {
  menuItems: MenuItem[];
  isVisible: boolean;
  onItemClick: () => void;
};

function isActive(item: MenuItem, pathName: string): boolean {
  if (item.activePattern) {
    return item.activePattern.test(pathName);
  }
  return pathName === item.path;
}

export const MobileMenu = forwardRef<HTMLDivElement, MobileMenuProps>(
  ({ menuItems, isVisible, onItemClick }, ref) => {
    const pathName = usePathname();

    return (
      <div
        id="mobile-menu"
        ref={ref}
        className={`sm:hidden ${styles.mobileMenu} ${
          isVisible ? styles.expanded : ""
        }`}
        aria-hidden={!isVisible}
      >
        <div className="px-4 pt-2 pb-3 space-y-1">
          {menuItems.map((item) => (
            <Link
              key={item.path}
              href={item.path}
              className={`${styles.navItem} ${
                isActive(item, pathName) ? styles["active"] : ""
              }`}
              onClick={onItemClick}
            >
              {item.label}
            </Link>
          ))}
        </div>
      </div>
    );
  }
);

MobileMenu.displayName = "MobileMenu";