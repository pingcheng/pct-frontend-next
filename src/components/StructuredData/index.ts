export * from "./StructuredData";
export * from "./PersonStructuredDataProvider";
