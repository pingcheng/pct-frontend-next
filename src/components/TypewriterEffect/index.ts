export { TypewriterEffect } from "./TypewriterEffect";
export { CyclingTypewriterEffect } from "./CyclingTypewriterEffect";