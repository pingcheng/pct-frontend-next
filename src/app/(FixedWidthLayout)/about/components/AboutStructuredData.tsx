import {
  BreadcrumbListStructuredData,
  ProfessionalServiceStructuredData,
} from "@/components/StructuredData/StructuredData";
import { PersonStructuredDataProvider } from "@/components/StructuredData/PersonStructuredDataProvider";
import { profile } from "@/data/profile";
import { backendStack, devOpsStack, frontendStack, legacyStack } from "@/data/skills";
import { URLS } from "@/constants/urls";

export function AboutStructuredData() {
  const skillsArray = [
    ...backendStack,
    ...frontendStack,
    ...devOpsStack,
    ...legacyStack,
  ];

  const breadcrumbItems = [
    { name: "Home", url: URLS.WEBSITE.BASE },
    { name: "About", url: URLS.WEBSITE.ABOUT },
  ];

  return (
    <>
      <PersonStructuredDataProvider 
        knowsAbout={skillsArray}
        includeEmail={true}
      />
      <BreadcrumbListStructuredData items={breadcrumbItems} />
      <ProfessionalServiceStructuredData
        name="Software Development Services"
        description="Full-stack web development, cloud solutions, and enterprise software development services specializing in Node.js, TypeScript, PHP, Java, and modern web technologies."
        provider={{
          name: profile.fullName,
          url: URLS.WEBSITE.BASE,
        }}
        areaServed="Melbourne, Australia"
        serviceType="Software Development"
      />
    </>
  );
}