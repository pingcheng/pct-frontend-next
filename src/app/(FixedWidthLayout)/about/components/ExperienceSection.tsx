import { Heading } from "@/components/Heading";
import { workExperiences } from "@/data/work-experience";
import { FaLinkedin } from "react-icons/fa";
import { getDurationString } from "@/utils/getDurationString";

function renderDescriptionLine(line: string, descIdx: number) {
  if (line === ":line-break:") {
    return <div key={descIdx} className="h-1"></div>;
  }
  
  if (line.startsWith("- ")) {
    return (
      <div key={descIdx} className="text-xs text-gray-600 dark:text-gray-400 leading-relaxed ml-3">
        • {line.substring(2)}
      </div>
    );
  }
  
  if (line.endsWith(":")) {
    return (
      <div key={descIdx} className="text-xs font-medium text-gray-700 dark:text-gray-300 mt-2 mb-1">
        {line}
      </div>
    );
  }
  
  return (
    <div key={descIdx} className="text-xs text-gray-600 dark:text-gray-400 leading-relaxed">
      {line}
    </div>
  );
}

export function ExperienceSection() {
  return (
    <section className="card p-6 rounded-lg w-full animate-fade-in animate-delay-1300">
      <div className="mb-6 animate-fade-in animate-delay-1400">
        <Heading text="Professional Experience" align="center" />
      </div>
      <div className="space-y-6">
        {workExperiences.map((companyItem, companyIdx) => (
          <div key={companyIdx} className={`border-l-2 border-gray-200 dark:border-gray-700 pl-4 animate-slide-in-left animate-delay-${1500 + (companyIdx * 200)}`}>
            <div className="mb-4">
              <div className="flex items-center gap-2 mb-1">
                <h3 className="text-base font-semibold text-gray-900 dark:text-gray-100">
                  {companyItem.company.name}
                </h3>
                {companyItem.company.linkedInUrl && (
                  <a
                    href={companyItem.company.linkedInUrl}
                    target="_blank"
                    rel="noopener noreferrer"
                    className="text-gray-500 hover:text-gray-700 dark:hover:text-gray-300 transition-colors"
                    aria-label={`${companyItem.company.name} LinkedIn profile`}
                  >
                    <FaLinkedin size={14} aria-hidden="true" />
                  </a>
                )}
              </div>

              <div className="space-y-3">
                {companyItem.positions.map((item, idx) => (
                  <div key={idx}>
                    <h4 className="font-medium text-gray-800 dark:text-gray-200 text-sm mb-1">
                      {item.position}
                    </h4>
                    <div className="text-xs text-gray-500 dark:text-gray-400 mb-2">
                      {item.startDate} - {item.endDate} · {getDurationString(item.startDate, item.endDate)}
                    </div>
                    {item.description.length > 0 && (
                      <div className="space-y-1">
                        {item.description.map((line, descIdx) => renderDescriptionLine(line, descIdx))}
                      </div>
                    )}
                  </div>
                ))}
              </div>
            </div>
          </div>
        ))}
      </div>
    </section>
  );
}