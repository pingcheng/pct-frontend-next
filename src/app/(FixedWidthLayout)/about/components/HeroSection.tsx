import { Avatar } from "@/components/Avatar";
import { Heading } from "@/components/Heading";
import { profile, socialUrls } from "@/data/profile";
import { FaLinkedin, FaGithub } from "react-icons/fa";

export function HeroSection() {
  const links = [
    { label: "github", url: socialUrls.github, text: "GitHub", ariaLabel: "Visit my GitHub profile" },
    { label: "linkedin", url: socialUrls.linkedin, text: "LinkedIn", ariaLabel: "Visit my LinkedIn profile" },
    { label: "email", url: `mailto:${profile.email}`, text: profile.email, ariaLabel: `Send email to ${profile.email}` },
  ];

  return (
    <section className="card p-8 rounded-lg w-full mb-8 animate-fade-in-scale">
      <div className="flex flex-col lg:flex-row items-center gap-8">
        <div className="flex-shrink-0 animate-fade-in animate-delay-200">
          <div className="shadow-lg" style={{ borderRadius: "50%" }}>
            <Avatar width={120} height={120} alt="Ping Cheng profile photo" />
          </div>
        </div>
        <div className="flex-1 text-center lg:text-left">
          <div className="mb-3 animate-fade-in animate-delay-300">
            <Heading text="Ping Cheng" align="left" />
          </div>
          <h2 className="text-lg font-medium text-gray-800 dark:text-gray-200 mb-4 animate-fade-in animate-delay-400">
            Staff Engineer at REA Group
          </h2>
          <p className="text-gray-600 dark:text-gray-400 leading-relaxed mb-6 animate-fade-in animate-delay-500">
            Experienced full-stack engineer specializing in modern web technologies, cloud solutions, and enterprise software development.
            Passionate about building scalable systems and leading technical initiatives that drive business value.
          </p>
          <div className="flex flex-wrap gap-3 justify-center lg:justify-start animate-fade-in animate-delay-600">
            {links.map((link) => (
              <a
                key={link.label}
                href={link.url}
                target="_blank"
                rel="noopener noreferrer"
                className="inline-flex items-center gap-2 px-3 py-1.5 text-gray-700 dark:text-gray-300 border border-gray-300 dark:border-gray-600 rounded-md hover:bg-gray-50 dark:hover:bg-gray-700 transition-colors"
                aria-label={link.ariaLabel}
              >
                {link.label === 'github' && <FaGithub size={14} />}
                {link.label === 'linkedin' && <FaLinkedin size={14} />}
                {link.label === 'email' && '✉️'}
                <span className="text-xs">{link.text}</span>
              </a>
            ))}
          </div>
        </div>
      </div>
    </section>
  );
}