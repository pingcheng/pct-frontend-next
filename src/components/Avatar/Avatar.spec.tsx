import { render, screen, fireEvent } from "@testing-library/react";
import { Avatar } from "@/components/Avatar";

describe("test <Avatar />", () => {
  it("should match snapshot", () => {
    const { container } = render(<Avatar width={100} height={100} />);
    expect(container).toMatchSnapshot();
  });

  it("should have image defined", () => {
    render(<Avatar width={100} height={100} />);
    expect(screen.getByRole("img")).toBeInTheDocument();
  });

  it("should have correct width and height", () => {
    render(<Avatar width={100} height={100} />);
    expect(screen.getByRole("img")).toHaveAttribute("width", "100");
    expect(screen.getByRole("img")).toHaveAttribute("height", "100");
  });

  it("should handle image error", () => {
    render(<Avatar width={100} height={100} />);
    const img = screen.getByRole("img");
    
    fireEvent.error(img);
    
    // After error, should show fallback avatar
    expect(screen.getByText("No Image")).toBeInTheDocument();
  });
});
