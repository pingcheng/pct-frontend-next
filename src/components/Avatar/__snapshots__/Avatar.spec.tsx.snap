// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <Avatar /> should match snapshot 1`] = `
<div>
  <div
    class="max-w-full "
    style="width: 100px; height: 100px;"
  >
    <img
      alt="Ping Cheng profile photo"
      data-nimg="1"
      decoding="async"
      height="100"
      loading="lazy"
      src="http://localhost/_next/image?url=https%3A%2F%2Favatars0.githubusercontent.com%2Fu%2F8447539&w=256&q=75"
      srcset="/_next/image?url=https%3A%2F%2Favatars0.githubusercontent.com%2Fu%2F8447539&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Favatars0.githubusercontent.com%2Fu%2F8447539&w=256&q=75 2x"
      style="color: transparent; border-radius: 50%;"
      width="100"
    />
  </div>
</div>
`;
